examples/raytracing/example_raytracing.cpp

```diff
@@ -119,83 +223,179 @@ int main(int argc, char *argv[])
+    Kokkos::View<int *> offsets("offsets", 0);
+    bvh.query(exec_space, IntersectsBased::Rays<MemorySpace>{rays},
+              IntersectsBased::AccumRaySphereOptDist<MemorySpace>{boxes},
+              values, offsets);
+
+    Kokkos::View<IntersectsBased::IntersectedCellForSorting *, MemorySpace>
+        sort_array(Kokkos::view_alloc(exec_space, Kokkos::WithoutInitializing,
+                                      "sort_array"),
+                   values.size());
+    Kokkos::parallel_for(
+        "copy sort_array",
+        Kokkos::RangePolicy<ExecutionSpace>(exec_space, 0, values.size()),
+        KOKKOS_LAMBDA(int i) { sort_array(i) = values(i); });
+    Kokkos::Profiling::pushRegion("sorting");
+#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP) ||               \
+    defined(KOKKOS_ENABLE_SYCL)
+    auto permutation = ArborX::Details::sortObjects(exec_space, sort_array);
+#else
+    Kokkos::View<int *> permutation(
+        Kokkos::view_alloc("permutation", Kokkos::WithoutInitializing),
+        sort_array.size());
+    std::iota(permutation.data(), permutation.data() + sort_array.size(), 0);
+    std::sort(permutation.data(), permutation.data() + sort_array.size(),
+              [&](int const &a, int const &b) {
+                return (sort_array(a) < sort_array(b));
+              });
+#endif
+    Kokkos::Profiling::popRegion();
+
+    energy_intersects =
+        Kokkos::View<float *, MemorySpace>("energy_intersects", num_boxes);
+    Kokkos::parallel_for(
+        "deposit_energy",
+        Kokkos::RangePolicy<ExecutionSpace>(exec_space, 0,
+                                            num_rays * num_boxes),
+        KOKKOS_LAMBDA(int i) {
+          using Kokkos::Experimental::expm1;
+          float ray_energy = (4000. * dx * dy * dz) / num_rays;
+          for (int j = offsets(i); j < offsets(i + 1); ++j)
+          {
+            float const energy_deposited =
+                -ray_energy *
+                expm1(-values(permutation(j)).optical_path_length);
+            ray_energy += energy_deposited;
+            Kokkos::atomic_add(&energy_intersects(values(permutation(j)).cid),
+                               energy_deposited);
+          }
+        });
+    Kokkos::Profiling::popRegion();
+  }
 
-  printf("time          : %.3f   [%.3fM ray/sec]\n", time,
-         num_rays / (1000000 * time));
-  printf("ray avg       : %.3f\n", accumulator_avg);
+  // Now check that the results we got are the same apart from numerical errors
+  // introduced by depositing energy to a particular cell from separate rays
+  // in different order.
+  int n_errors = 0;
+  float rel_tol = 1.e-5;
+  Kokkos::parallel_reduce(
+      "compare", Kokkos::RangePolicy<ExecutionSpace>(exec_space, 0, num_boxes),
+      KOKKOS_LAMBDA(int i, int &error) {
+        using Kokkos::Experimental::fabs;
+        float const rel_error =
+            (energy_intersects(i) == 0.f)
+                ? 0.f
```

Comment:
Not sure about this. If `energy_intersects(i)` is 0, and `energy_ordered_intersects(i)` is not, why is the relative error 0?

Reply:
Changed.